### PR TITLE
Allow large JSON numbers in provider2/upgradeState

### PR DESCRIPTION
**Pull request by iwahbe**
Stacked on top of #1998

---

This PR allows provider2.upgradeState to handle 64 bit integer numbers. The substance of the change is https://github.com/pulumi/terraform-plugin-sdk/pull/39. 

The only change here is a test.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,7 +47,7 @@ tidy::
 # on windows.
 pin_upstream_sdk: UpstreamPluginSDK=github.com/hashicorp/terraform-plugin-sdk/v2
 pin_upstream_sdk: OurPluginSDK=github.com/pulumi/terraform-plugin-sdk/v2
-pin_upstream_sdk: PluginSDKVersion=v2.0.0-20230912190043-e6d96b3b8f7e
+pin_upstream_sdk: PluginSDKVersion=v2.0.0-20240520223432-0c0bf0d65f10
```
Comment (VenelinMartinov) [inline Makefile:50]: Can you raise a PR there as well so the changes can be reviewed? We don't need to merge it

File: pkg/tfshim/tfplugin5/provider.go [hunk 2]
```diff
@@ -403,7 +403,7 @@ func (p *provider) Apply(
 		return nil, err
 	}
 
-	resp, err := p.client.ApplyResourceChange(context.TODO(), &proto.ApplyResourceChange_Request{
+	resp, err := p.client.ApplyResourceChange(ctx, &proto.ApplyResourceChange_Request{
```
Comment (VenelinMartinov) [inline pkg/tfshim/tfplugin5/provider.go:406]: Appreciate you cleaning up here but these seem unrelated to the PR, should we separate them?

File: pkg/tfshim/sdk-v2/provider2.go [hunk 3]
```diff
@@ -326,11 +327,7 @@ func (p *planResourceChangeImpl) upgradeState(
 
 	ty := res.CoreConfigSchema().ImpliedType()
 
-	jsonIsh, err := schema.StateValueToJSONMap(state.stateValue, ty)
-	if err != nil {
-		return nil, err
-	}
-	jsonBytes, err := json.Marshal(jsonIsh)
+	jsonBytes, err := ctyjson.Marshal(state.stateValue, ty)
```
Comment (VenelinMartinov) [inline pkg/tfshim/sdk-v2/provider2.go:330]: Can you explain the change here - seems like this is cutting out a bunch of code, which looks good but what was the issue with the previous code?

Comment (VenelinMartinov) [reply pkg/tfshim/sdk-v2/provider2.go:330]: Ah, saw your commit message about it, thanks!

**Comment by t0yv0** [comment]
Help us review the SDKv2 change please. I understand you're removing an old modification from Pat, how can we confirm that the original reason for that modification to exist is removed now, that is we are not introducing bugs? Sounded like it was about custom timeout support? Any more information? Can summon Pat here.

File: Makefile (hunk 1 above)
Comment (iwahbe) [reply Makefile:50]: I have opened https://github.com/pulumi/terraform-plugin-sdk/pull/39 to track the pulumi/terraform-plugin-sdk change.

It looks like the original issues are resolved. We have tests for resource timeouts:

- https://github.com/pulumi/pulumi-terraform-bridge/blob/a427483773c5ca2d8e9b235952276c63494294bb/pkg/tfbridge/schema_test.go#L695-L696
- https://github.com/pulumi/pulumi-terraform-bridge/blob/a427483773c5ca2d8e9b235952276c63494294bb/pkg/tfbridge/provider_test.go#L3735-L3736
- https://github.com/pulumi/pulumi-terraform-bridge/blob/a427483773c5ca2d8e9b235952276c63494294bb/pkg/tfshim/tfplugin5/provider_test.go#L1069

File: pkg/tfshim/tfplugin5/provider.go (hunk 2 above)
Comment (iwahbe) [reply pkg/tfshim/tfplugin5/provider.go]: Yeah. I removed them from this PR and I'll open a new PR.

**Comment by iwahbe** [comment]
> Help us review the SDKv2 change please. I understand you're removing an old modification from Pat, how can we confirm that the original reason for that modification to exist is removed now, that is we are not introducing bugs? Sounded like it was about custom timeout support? Any more information? Can summon Pat here.

I've reached out to @pgavlin for context (it's 4 years ago though). The commit message mentions custom timeouts, and we have tests in place for them (https://github.com/pulumi/pulumi-terraform-bridge/pull/2001#discussion_r1608753292).

**Comment by t0yv0** [comment]
Let's ship this. Please include clear messaging in the release notes about how to update to the new revision of the forked SDK. It's interesting because if providers don't build-break I assume most of the ecosystem will fail to take the update, without a reminder of a need to do so..

**Comment by pgavlin** [comment]
(sent this to @iwahbe in Slack as well, adding here for posterity)

I'm having difficulty remembering exactly what the issue was here. Best guesses are:

* That these lines were a problem, as they unconditionally use `json.Number`: https://github.com/pulumi/terraform-plugin-sdk/commit/53f910a1ea61564de1cae8f2e71d64ce4ecaf655#diff-cb044934d551a7bd462735476d13f6a21549a098d631f9ba5a2be806657d333fL563-R887
* Something in the bridge was choking on `json.Number`
* Some incompatibility in the way we translate between Pulumi state and TF state

**Review by t0yv0 — APPROVED** [review]
